### ISSUE TITLE
Spurious error for partial unknown

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#6763]
The following was formerly detected properly under strict, and did not complain about partially unknown for functions and classes which were yet to be used and where that information would come in at use.

```python
import itertools
from tqdm.contrib.itertools import product as tqdm_product

# later in the file, conditionally displaying progress via tqdm depending on various criteria. `use_tqdm` is a bool
gen_func = tqdm_product if use_tqdm else itertools.product
for candidates in in gen_func(*filter(None, cans)):
    ...
```

As of pylance Language Server version: 2023.12.1, the following error is shown for this:

```
Type of "product" is partially unknown
  Type of "product" is "type[product[Unknown]]"
```

The equivalent code below does not error on the current version

```py
filt = filter(None, cans)
gen = tqdm_product(*filt) if use_tqdm else itertools.product(*filt)
for candidates in gen:
    ...
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

via pylance in VSC, pylance Language Server version: 2023.12.1

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

There have been no changes in this behavior recently. I confirmed that the behavior is the same across many versions of pyright and pylance over the past three months. Specifically, I installed the version of pylance prior to 2023.12.1 (which was 2023.11.10) and confirmed that it generates the same errors.

I'm guessing that you've changed your settings and/or installed a different version of the `tqdm` library.

**Comment by erictraut** [comment]
Also, please double check the code that you've posted above. The first code snippet contains a syntax error and the second code snippet references symbols that aren't defined, so maybe it wasn't quite the code that you had intended to post.

**Comment by mikeshardmind** [comment]
> There have been no changes in this behavior recently. I confirmed that the behavior is the same across many versions of pyright and pylance over the past three months. Specifically, I installed the version of pylance prior to 2023.12.1 (which was 2023.11.10) and confirmed that it generates the same errors.
> 
> I'm guessing that you've changed your settings and/or installed a different version of the tqdm library

@erictraut The versions in use have not changed, nor have the settings (which are configured via pyproject.toml, committed to version control, and settings for pyright have not been touched since the project's first commit.)

I'm not sure what you mean that pyright is working as intended here or why the version of tqdm should matter, the error is on itertools.product, not on tqdm's wrapping of it

> Also, please double check the code that you've posted above. The first code snippet contains a syntax error and the second code snippet references symbols that aren't defined, so maybe it wasn't quite the code that you had intended to post.

Sorry, removing some surrounding irrelevant code led to an issue with that, below is a true minimum reproduction

```py

import itertools
from typing import reveal_type


def example(condition: bool = True) -> None:

    s = itertools.product if condition else None
    if s:
        for a, b in s((1,2,3), ("a","b","c")):
            reveal_type(a)
            reveal_type(b)
```
```
Type of "s" is partially unknown
  Type of "s" is "type[product[Unknown]] | None"
Type of "a" is "Literal[1, 2, 3]"
Type of "b" is "Literal['a', 'b', 'c']"
```

I've completely removed the tqdm component from this. pyright is eagerly complaining that it doesn't know that itertools.product is generic over while it is still a type, not an instance of a type.

**Comment by erictraut** [comment]
Thanks for the updated code. I'm able to repro the issue you're seeing. This was fixed in version 1.1.339 of pyright. It should be included in the next production release of pylance. If you switch to the latest prerelease version of pylance, you'll see that the error is no longer reported.
